File: lib/src/state/event.ts
import * as borsh from "borsh";
import { Buffer } from "buffer";
import { PublicKey } from "@solana/web3.js";

export class Event {

    id: number;
    authority: PublicKey;

    constructor(props: {
        id: number,
        authority: PublicKey,
    }) {
        this.id = props.id;
        this.authority = props.authority;
    }

    toBuffer() { 
        return Buffer.from(borsh.serialize(EventSchema, this)) 
    }
    
    static fromBuffer(buffer: Buffer) {
        const data = borsh.deserialize(EventSchema, Event, buffer);
        return new Event({
            id: data.id,
            authority: new PublicKey(data.authority),
        });
    }
}

export const EventSchema = new Map([
    [ Event, { 
        kind: 'struct', 
        fields: [ 
            ['id', 'u16'],
            ['authority', [32]],
        ],
    }]
]);

export class EventMetadata {

    authority: PublicKey;
    event: PublicKey;
    title: string;
    description: string;
    location: string;
    host: string;
    date: string;
    uri: string;

    constructor(props: {
        authority: PublicKey,
        event: PublicKey,
        title: string,
        description: string,
        location: string,
        host: string,
        date: string,
        uri: string,
    }) {
        this.authority = props.authority;
        this.event = props.event;
        this.title = props.title;
        this.description = props.description;
        this.location = props.location;
        this.host = props.host;
        this.date = props.date;
        this.uri = props.uri;
    }

    toBuffer() { 
        return Buffer.from(borsh.serialize(EventMetadataSchema, this)) 
    }
    
    static fromBuffer(buffer: Buffer) {
        const data = borsh.deserialize(EventMetadataSchema, EventMetadata, buffer);
        return new EventMetadata({
            authority: new PublicKey(data.authority),
            event: new PublicKey(data.event),
            title: data.title,
            description: data.description,
            location: data.location,
            host: data.host,
            date: data.date,
            uri: data.uri,
        });
    }
}

export const EventMetadataSchema = new Map([
    [ EventMetadata, { 
        kind: 'struct', 
        fields: [ 
            ['authority', [32]],
            ['event', [32]],
            ['title', 'string'],
            ['description', 'string'],
            ['location', 'string'],
            ['host', 'string'],
            ['date', 'string'],
            ['uri', 'string'],
        ],
    }]
]);

export class EventCounter {

    authority: PublicKey;
    count: number;

    constructor(props: {
        authority: PublicKey,
        count: number,
    }) {
        this.authority = props.authority;
        this.count = props.count;
    }

    toBuffer() { 
        return Buffer.from(borsh.serialize(EventCounterSchema, this)) 
    }
    
    static fromBuffer(buffer: Buffer) {
        const data = borsh.deserialize(EventCounterSchema, EventCounter, buffer);
        return new EventCounter({
            count: data.count,
            authority: new PublicKey(data.authority),
        })
    }
}

export const EventCounterSchema = new Map([
    [ EventCounter, { 
        kind: 'struct', 
        fields: [ 
            ['authority', [32]],
            ['count', 'u16'],
        ],
    }]
]);

File: program/src/state/event.rs
use borsh::{ BorshDeserialize, BorshSerialize };
use solana_program::pubkey::Pubkey;

use crate::state::{ 
    PrestigeDataAccount,
    PrestigeRoleAccount,
};

#[derive(BorshDeserialize, BorshSerialize)]
pub struct Event {
    pub id: u16,
    pub authority: Pubkey,
}

impl PrestigeDataAccount for Event {

    const SEED_PREFIX: &'static str = "event";

    fn bump(&self, program_id: &Pubkey) -> u8 {
        Pubkey::find_program_address(
            &[
                Event::SEED_PREFIX.as_bytes().as_ref(),
                self.authority.as_ref(),
                self.id.to_le_bytes().as_ref(),
            ],
            program_id
        ).1
    }

    fn is_valid_pda(&self, program_id: &Pubkey, address: &Pubkey) {
        assert!(
            &Pubkey::find_program_address(
                &[
                    Event::SEED_PREFIX.as_bytes().as_ref(),
                    self.authority.as_ref(),
                    self.id.to_le_bytes().as_ref(),
                ],
                program_id
            ).0 == address
        )
    }
}

impl PrestigeRoleAccount for Event {
    fn is_correct_authority(&self, address: &Pubkey) {
        assert!(&self.authority == address)
    }
}

#[derive(BorshDeserialize, BorshSerialize)]
pub struct EventMetadata {
    pub authority: Pubkey,
    pub event: Pubkey,
    pub title: String,
    pub description: String,
    pub location: String,
    pub date: String,
    pub host: String,
    pub uri: String,
}

impl PrestigeDataAccount for EventMetadata {

    const SEED_PREFIX: &'static str = "event_metadata";

    fn bump(&self, program_id: &Pubkey) -> u8 {
        Pubkey::find_program_address(
            &[
                EventMetadata::SEED_PREFIX.as_bytes().as_ref(),
                self.event.as_ref(),
            ],
            program_id
        ).1
    }

    fn is_valid_pda(&self, program_id: &Pubkey, address: &Pubkey) {
        assert!(
            &Pubkey::find_program_address(
                &[
                    EventMetadata::SEED_PREFIX.as_bytes().as_ref(),
                    self.event.as_ref(),
                ],
                program_id
            ).0 == address
        )
    }
}

impl PrestigeRoleAccount for EventMetadata {
    fn is_correct_authority(&self, address: &Pubkey) {
        assert!(&self.authority == address)
    }
}

#[derive(BorshDeserialize, BorshSerialize)]
pub struct EventCounter {
    pub authority: Pubkey,
    pub count: u16,
}

impl PrestigeDataAccount for EventCounter {

    const SEED_PREFIX: &'static str = "event_counter";

    fn bump(&self, program_id: &Pubkey) -> u8 {
        Pubkey::find_program_address(
            &[
                EventCounter::SEED_PREFIX.as_bytes().as_ref(),
                self.authority.as_ref(),
            ],
            program_id
        ).1
    }

    fn is_valid_pda(&self, program_id: &Pubkey, address: &Pubkey) {
        assert!(
            &Pubkey::find_program_address(
                &[
                    EventCounter::SEED_PREFIX.as_bytes().as_ref(),
                    self.authority.as_ref(),
                ],
                program_id
            ).0 == address
        )
    }
}